app/hooks/useJsonSchemaDiff.ts

```diff
@@ -0,0 +1,92 @@
+import diff from 'deep-diff';
+
+const addObjectToDiffMap = (path, lhs, rhs, idDiffMap) => {
+  if (typeof lhs !== 'object' && typeof rhs !== 'object') {
+    idDiffMap[path] = {lhs: lhs ?? null, rhs: rhs ?? null};
+    return;
+  }
+  if (typeof lhs === 'object' || typeof rhs === 'object') {
+    const keys = new Set<string>();
+
+    if (lhs) Object.keys(lhs).forEach((k) => keys.add(k));
+    if (rhs) Object.keys(rhs).forEach((k) => keys.add(k));
+
+    keys.forEach((k) =>
+      addObjectToDiffMap(`${path}_${k}`, lhs?.[k], rhs?.[k], idDiffMap)
+    );
+  }
+};
+
+/**
+ *
+ * @param formData the form data to display when not diffing
+ * @param showDiff whether or not the diff should be performed
+ * @param formIdPrefix
+ * @param lhs
+ * @param rhs
+ * @returns An object with the formData to display, and the idDiffMap, containing the field-level diffs
+ */
+const useJsonSchemaDiff = (
+  formData: any,
+  showDiff: boolean,
+  formIdPrefix: string,
+  lhs?: any,
+  rhs?: any
+) => {
+  if (!showDiff) return {formData};
```

Comment:
I would handle the diffing toggle outside this function, make that the responsibility of the caller

Reply:
That would break the first rule of hooks: https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

Technically this doesn't need to be a hook though, since it doesn't use any state/effect

Reply:
Ah, I didn't see that part about it being a hook.
Maybe the hook could just be the toggle - but the function could live in a library by itself?